tools/translate_annotations.py

```diff
@@ -0,0 +1,172 @@
+        hunk_match = re.match(hunk_pattern, line)
+        if name:
+            assert not name_match
+            if hunk_match:
+                old_start, old_count, new_start, new_count = hunk_match.groups()
+                hunks.append({
+                    'old_start': int(old_start),
+                    'old_count': int(old_count or '1'),
+                    'new_start': int(new_start),
+                    'new_count': int(new_count or '1'),
+                })
+        else:
+            assert not hunk_match
+            if name_match:
+                name, = name_match.groups()
+    assert name
+    return {
+        'old_filename': name,
+        'hunks': hunks,
+    }
+
+
+T = TypeVar('T')
+U = TypeVar('U')
+
+
+# https://gist.github.com/ericremoreynolds/2d80300dabc70eebc790
+class KeyifyList(Generic[T, U]):
+    def __init__(self, inner: List[T], key: Callable[[T], U]) -> None:
+        self.inner = inner
+        self.key = key
+
+    def __len__(self) -> int:
+        return len(self.inner)
+
+    def __getitem__(self, k: int) -> U:
+        return self.key(self.inner[k])
+
+
+def translate(diff: Diff, line_number: int) -> Optional[int]:
+    if line_number < 1:
+        return None
+
+    hunks = diff['hunks']
+    if not hunks:
+        return line_number
+
+    keyified = KeyifyList(
+        hunks,
+        lambda hunk: hunk['new_start'] + (0 if hunk['new_count'] > 0 else 1)
+    )
+    i = bisect_right(cast(Sequence[int], keyified), line_number)
+    if i < 1:
+        return line_number
+
+    hunk = hunks[i - 1]
+    d = line_number - (hunk['new_start'] + (hunk['new_count'] or 1))
+    return None if d < 0 else hunk['old_start'] + (hunk['old_count'] or 1) + d
+
+
+# we use camelCase here because this will be output as JSON and so the
+# field names need to match the group names from here:
+# https://github.com/pytorch/add-annotations-github-action/blob/3ab7d7345209f5299d53303f7aaca7d3bc09e250/action.yml#L23
+class Annotation(TypedDict):
+    filename: str
+    lineNumber: int
+    columnNumber: int
+    errorCode: str
+    errorDesc: str
+
+
+def parse_annotation(regex: Pattern[str], line: str) -> Optional[Annotation]:
+    m = re.match(regex, line)
+    if m:
+        try:
+            line_number = int(m.group('lineNumber'))
+            column_number = int(m.group('columnNumber'))
+        except ValueError:
+            return None
+        return {
+            'filename': m.group('filename'),
+            'lineNumber': line_number,
+            'columnNumber': column_number,
+            'errorCode': m.group('errorCode'),
+            'errorDesc': m.group('errorDesc'),
+        }
+    else:
+        return None
+
+
+def translate_all(
+    *,
+    lines: List[str],
+    regex: Pattern[str],
+    commit: str
+) -> List[Annotation]:
+    ann_dict: DefaultDict[str, List[Annotation]] = defaultdict(list)
+    for line in lines:
+        annotation = parse_annotation(regex, line)
+        if annotation:
```

Comment:
Not sure what is right convention here, but `if annotation` will fail if `annotation` is an empty string
```suggestion
        if annotation is not None:
```

Reply:
In this case since `annotation` is of type `Optional[Annotation]` it's impossible for it to be an empty string (or empty dict etc) but I can change it to `is not None` for clarity in any case